### PR TITLE
[build-properties] Expand extraMavenRepos support

### PR DESCRIPTION
**Pull request by bpeltonc**
- Allow users to specify authentication scheme and credentials for repositories added to the project via `android.extraMavenRepos` per Gradle/Maven documentation.

# Why

Support was added in Expo Build Properties 0.8.0 to allow users to specify additional Maven repositories to be included with an Android build, but this initial implementation is somewhat naive and only offers the user a small portion of the functionality Gradle and Maven offers, namely only allowing users to add repositories that don't require credentials. Users who would like to include Maven repositories that do require credentials are precluded from using Expo Build Properties to add them and would appreciate the ability to use the `extraMavenRepos` support to include these repos in their builds as well.

# How

1. I provided TypeScript definitions for Maven repository-related objects conforming to the protocols outlined in the [Gradle documentation](https://docs.gradle.org/current/userguide/declaring_repositories.html#sec:authentication_schemes) and modified the `extraMavenRepos` definition to accept an array of either these objects or strings, providing backwards-compatibility to consumers of this library who upgrade to this version and averting introducing breaking changes.
2. I adjusted the AJV schema to properly validate user-provided values against the newly-defined types.
3. I made updates to `expo-modules-autolinking` to accept the new properties and include them in the autolinking script.
4. I wrote automated tests to ensure user-provided values are properly validated against the Gradle protocols as expressed in the TypeScript defintions and to ensure those values were passed in the expected data structure to the `expo-modules-autolinking` package.

# Test Plan

One of the projects I work on specifies an extra Maven repository that requires a credentialed login. I specified a object in the `android.extraMavenRepos` section of my `app.config.ts` file that included a url, credentials object, and authentication scheme. Then I ran `npx expo prebuild --clean` and `npx expo android:run`. With correct credentials, the app builds and runs successfully, and the feature provided by the repo is usable from within the Expo application.

For reviewers of this PR, the main obstacle in testing will likely be finding a Maven repository that needs credentials for access, and obtaining valid credentials. I wish I could provide the credentials I used to test, but they are, of course, sensitive. Hopefully, we can collaborate on devising an easier way to test this PR.

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Review by lukmccall — APPROVED** [review]
It's looking good. Let's wait for @Kudo, but if you ask me, we can merge it ;)

**Comment by bpeltonc** [comment]
@Kudo Would you mind reviewing/merging this when you have the chance?

File: packages/expo-build-properties/src/pluginConfig.ts
Comment (Kudo) [inline packages/expo-build-properties/src/pluginConfig.ts:172]:
```suggestion
  authentication?: 'basic' | 'digest' | 'header';
```

that's a very nit comment. internally we prefer not to use typescript enum. wdyt if we could replace with string union here?

**Review by Kudo — COMMENTED** [review]
sorry i missed this and thanks for your reminder as well as the great pr! i like the pr much for having great test coverage / comment / and nice code. i'm just leave a comment for not using typescript enum. let me know how do you feel for that.

File: packages/expo-build-properties/src/pluginConfig.ts
Comment (bpeltonc) [reply packages/expo-build-properties/src/pluginConfig.ts]: No problem here. I can understand why some prefer the string union. This change is pushed.

File: packages/expo-modules-autolinking/src/types.ts
Comment (bpeltonc) [reply packages/expo-modules-autolinking/src/types.ts]: Fixed that too

**Comment by Kudo** [comment]
sorry there're more 🤦‍♂️ i still dunno how to push directly to a pr where authors allowing maintainer to edit. please help me to edit the change again.

```diff
commit b119bc25b01cbde4cd459d248b9f91f1700f2c1e
diff --git a/packages/expo-modules-autolinking/build/types.d.ts b/packages/expo-modules-autolinking/build/types.d.ts
index 9146e0306c..587d9149c8 100644
Binary files a/packages/expo-modules-autolinking/build/types.d.ts and b/packages/expo-modules-autolinking/build/types.d.ts differ
diff --git a/packages/expo-modules-autolinking/build/types.js b/packages/expo-modules-autolinking/build/types.js
index 7e60b199d7..11e638d1ee 100644
Binary files a/packages/expo-modules-autolinking/build/types.js and b/packages/expo-modules-autolinking/build/types.js differ
diff --git a/packages/expo-modules-autolinking/build/types.js.map b/packages/expo-modules-autolinking/build/types.js.map
index 066206f0cc..9c759e3f5a 100644
Binary files a/packages/expo-modules-autolinking/build/types.js.map and b/packages/expo-modules-autolinking/build/types.js.map differ
diff --git a/packages/expo-modules-autolinking/src/autolinking/__tests__/extraDependencies-test.ts b/packages/expo-modules-autolinking/src/autolinking/__tests__/extraDependencies-test.ts
index 9473adbf54..7317bce7ba 100644
--- a/packages/expo-modules-autolinking/src/autolinking/__tests__/extraDependencies-test.ts
+++ b/packages/expo-modules-autolinking/src/autolinking/__tests__/extraDependencies-test.ts
@@ -1,6 +1,5 @@
 import { getConfig } from '@expo/config';
 
-import { AndroidMavenRepositoryAuthenticationScheme } from '../../types';
 import { getBuildPropertiesAsync, resolveExtraDependenciesAsync } from '../extraDependencies';
 
 jest.mock('@expo/config', () => ({
@@ -105,7 +104,7 @@ describe(resolveExtraDependenciesAsync, () => {
                       username: 'user',
                       password: 'password',
                     },
-                    authentication: AndroidMavenRepositoryAuthenticationScheme.BasicAuthentication,
+                    authentication: 'basic',
                   },
                 ],
               },
@@ -148,8 +147,7 @@ describe(resolveExtraDependenciesAsync, () => {
                       name: 'token',
                       value: 'some_token',
                     },
-                    authentication:
-                      AndroidMavenRepositoryAuthenticationScheme.HttpHeaderAuthentication,
+                    authentication: 'header',
                   },
                 ],
               },
@@ -192,7 +190,7 @@ describe(resolveExtraDependenciesAsync, () => {
                       username: 'user',
                       password: 'password',
                     },
-                    authentication: AndroidMavenRepositoryAuthenticationScheme.DigestAuthentication,
+                    authentication: 'digest',
                   },
                 ],
               },
diff --git a/packages/expo-modules-autolinking/src/types.ts b/packages/expo-modules-autolinking/src/types.ts
index fbee9d9816..44569036ec 100644
--- a/packages/expo-modules-autolinking/src/types.ts
+++ b/packages/expo-modules-autolinking/src/types.ts
@@ -228,12 +228,6 @@ export interface RawExpoModuleConfig {
   };
 }
 
-export enum AndroidMavenRepositoryAuthenticationScheme {
-  BasicAuthentication = 'basic',
-  DigestAuthentication = 'digest',
-  HttpHeaderAuthentication = 'header',
-}
-
 interface AndroidMavenRepositoryPasswordCredentials {
   username: string;
   password: string;

```

**Comment by bpeltonc** [comment]
Thanks for being thorough and catching all this. I'll get all this fixed and do a full sweep and test of my original PR before I push another hasty commit. I should have caught this.

**Comment by bpeltonc** [comment]
Removed all references to that enum and replaced them all with string unions, ran all the tests and build packagers again for both packages, and (after fixing a bug!) tested it again successfully linking a maven repo that requires basic authentication. I'm confident this is ready to go and addresses all your comments, @Kudo

**Review by Kudo — APPROVED** [review]
everything looks good and i'm going to merge this pr. thanks again for the fantastic contribution 👏

**Pull request merged**
